### PR TITLE
Store number of stack slots in vmThread.tempSlot for linkTo* calls

### PR DESCRIPTION
**Pull request by nbhuiyan**
linkTo* calls (linkToStatic, linkToSpecial, linkToVirtual and
linkToInterface) are also signature-polymorphic, requiring the JIT
to specify the number of stack slots used by the args
for the interpreter to locate the start of the args
on the stack. This is achieved by performing a store to
vmThread.tempSlot before the INL call, similar to how it was
done for invokeBasic. For invokeBasic, the value stored in
tempSlot is now the number of arg slots, and not the location of
the MH receiver object.

Signed-off-by: Nazim Bhuiyan <nubhuiyan@ibm.com>

Related: #11877

### CODE REVIEW

**Comment by nbhuiyan** [comment]
@babsingh @liqunl @0xdaryl FYI

File: runtime/compiler/codegen/J9CodeGenerator.cpp
Comment (liqunl) [inline runtime/compiler/codegen/J9CodeGenerator.cpp:1028]: The code was to pass location of the MethodHandle receiver for `invokeBasic`,  but `linkTo*` methods dont' have MethodHandle receiver. Maybe we should pass the number of argument slots (including receiver if exists), and VM can use this info differently for different methods. @babsingh If we were to write the number of argument slots to `tempSlot`, I believe VM has to modify your code in handling `invokeBasic` called from the JIT

Comment (babsingh) [reply runtime/compiler/codegen/J9CodeGenerator.cpp:1028]: Receiver is the first arg. The receiver's location can be derived knowing the number of argument slots. @fengxue-IS to confirm?

Comment (liqunl) [reply runtime/compiler/codegen/J9CodeGenerator.cpp:1028]: @nbhuiyan Can you change the code to write the number of argument slots to `tempSlot` instead of the location of the first argument? @babsingh VM has to modify the place that currently uses `tempSlot` 
https://github.com/eclipse/openj9/blob/3dcfa980467d67ac059129f7349fff6159469737/runtime/vm/BytecodeInterpreter.hpp#L8251

**Comment by babsingh** [comment]
@liqunl @nbhuiyan Does the JIT also set `currentThread->jitReturnAddress` before calling the `invokeBasic` or `linkTo*` INLs? It will be needed to build the native JIT resolve frame.

https://github.com/eclipse/openj9/blob/3dcfa980467d67ac059129f7349fff6159469737/runtime/codert_vm/cnathelp.cpp#L436-L437

**Comment by liqunl** [comment]
@babsingh I believe VM knows how to find it as the JIT calling VM is via j2itransition. @gacholio to confirm?
https://github.com/eclipse/openj9/blob/3dcfa980467d67ac059129f7349fff6159469737/runtime/vm/BytecodeInterpreter.hpp#L556

**Comment by nbhuiyan** [comment]
@liqunl requesting review.

**Comment by nbhuiyan** [comment]
@babsingh I have made the necessary changes for the invokeBasic handler

File: runtime/vm/BytecodeInterpreter.hpp
Comment (babsingh) [inline runtime/vm/BytecodeInterpreter.hpp:8251]:
```suggestion
			/* tempSlot contains the number of stack slots for the arguments, and the MH
			 * receiver is the first argument. 
			 */
```

**Comment by liqunl** [comment]
We should also compile time resolve linkTo methods since lowering only works for resolved calls.

File: runtime/vm/BytecodeInterpreter.hpp
Comment (babsingh) [reply runtime/vm/BytecodeInterpreter.hpp]: number of args != number of stack slots

File: runtime/compiler/codegen/J9CodeGenerator.cpp
Comment (liqunl) [inline runtime/compiler/codegen/J9CodeGenerator.cpp:785]: We should do the same change for 32bit platform

**Review by liqunl — APPROVED** [review]

**Comment by nbhuiyan** [comment]
@0xdaryl Requesting review

**Comment by gacholio** [comment]
> @babsingh I believe VM knows how to find it as the JIT calling VM is via j2itransition. @gacholio to confirm?

Confirmed - the VM knows how to find the return address in it's normal place (on stack or in a link register).

File: runtime/vm/BytecodeInterpreter.hpp
Comment (fjeremic) [inline runtime/vm/BytecodeInterpreter.hpp:8254]: Would you consider adding some documentation for the `tempSlot` field? I personally get worried when we start overloading uses of a "temp" field. We are also increasingly growing what kinds of things we store into that field. Questions I ask myself when encountering a use of `tempSlot`:

- What are all the uses of this field?
- How can I be sure that the value I store here will not be overwritten by some other piece of code which wants to use this same field?

Please see #12045 where we attempt to document an overloaded field like this. I don't expect us to document all uses, but if you start a similar template as seen in #12045 for this particular use, then we can set a precedent that all future uses should be documented as well.

Eventually we may not eliminate such uses, but at the very least there will be concentrated documentation right where the field is declared about all the uses of it. Hopefully this will help the reader understand the answers to the above two questions.

Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp:8254]: We've long discussed using individual fields for these things so that they could be somewhat self-documenting. The problem is that we need to keep the thread size down where possible, and that some of the fields that are accessed from the hand-coded assembly must be below a certain offset in order to use simple instruction forms on various processors. Similarly, some fields are accessed by the compiled code - those fields want to have small offsets to allow optimal encoding.

Comment (nbhuiyan) [reply runtime/vm/BytecodeInterpreter.hpp:8254]: @fjeremic, I agree that we should be documenting the usage of this field as much as we can. I have opened #12095 to track the work on documentation for this field.

Comment (fjeremic) [reply runtime/vm/BytecodeInterpreter.hpp:8254]: Thanks for opening that. I would like to see documentation added for the particular use case of this PR before we merge. Let's set the precedent and we can use the issue you opened to track addition of other documentation not related to this PR.

File: runtime/compiler/codegen/J9CodeGenerator.cpp
Comment (0xdaryl) [inline runtime/compiler/codegen/J9CodeGenerator.cpp:767]: Unless I'm mis-reading the logic, I think this code would be more readable (and you would help out the compiler to optimize it better) if you were to write it like this:
```
if (node->getOpCode().isCall() &&
    !node->getSymbol()->castToMethodSymbol()->isHelper())
   {
   TR::RecognizedMethod recognizedMethod = node->getSymbol()->castToMethodSymbol()->getRecognizedMethod();

   if (recognizedMethod == TR::java_lang_invoke_MethodHandle_invokeBasic ||
       recognizedMethod == TR::java_lang_invoke_MethodHandle_linkToStatic ||
       ...
      {
      ...
      }
```
Perhaps even consider a `switch` statement.

**Comment by 0xdaryl** [comment]
@gacholio : can you review/approve the changes to `runtime/vm/BytecodeInterpreter.hpp` please?

**Comment by gacholio** [comment]
@babsingh Have you tested with these changes?

**Comment by babsingh** [comment]
> Have you tested with these changes?

@gacholio I haven't run builds yet. You should be able to run PR builds with OJDK MHs enabled. `ojdk292` is suffixed to existing configs. Example commands:

```
jenkins compile xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29
jenkins compile winojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29
```

**Review by gacholio — APPROVED** [review]
I'll approve the interpreter changes - sounds like more JIT changes are requested, so I'm not going to run the builds.

**Review by 0xdaryl — APPROVED** [review]

**Comment by 0xdaryl** [comment]
jenkins compile xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

**Comment by 0xdaryl** [comment]
jenkins compile winojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

**Comment by 0xdaryl** [comment]
@babsingh : assuming the testing passes, how do you want to merge this?

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk16

**Comment by babsingh** [comment]
> how do you want to merge this?

@0xdaryl This PR is standalone and has no dependencies. It can be merged on its own without any complications.

This PR has been created to support the changes in #11877. So, this PR should only be merged once the changes in #11877 are finalized and working as expected. Once #11877 is ready, I will post a comment here indicating that this PR can be merged. Once this PR is merged, I will rebase before having #11877 merged.

**Review by fjeremic — APPROVED** [review]

**Comment by 0xdaryl** [comment]
Testing is successful and approvers have signed off.  I won't merge until I hear from @babsingh .

**Comment by babsingh** [comment]
@0xdaryl https://github.com/eclipse/openj9/pull/11877 (Build JIT Resolve Frame PR) has been tested, approved and ready to be merged. We can start by merging this PR. fyi @gacholio.

**Comment by 0xdaryl** [comment]
Merged.  @babsingh

**Pull request merged**
